### PR TITLE
TOTP: Add option to type TOTP

### PR DESCRIPTION
**Pull request by 0x0ptr**

### CODE REVIEW

File: source_code/main_mcu/src/LOGIC/logic_user.c [hunk 1]
```diff
@@ -1837,154 +1837,270 @@ RET_TYPE logic_user_ask_for_credentials_keyb_output(uint16_t parent_address, uin
...
                 return RETURN_OK;
             }
             else if (prompt_return == MINI_INPUT_RET_BACK)
             {
-                /* Check for no login */
-                if (temp_cnode.login[0] == 0)
+                /*
+                 * State is TOTP typing
+                 * If password is blank we go back 2 to login type prompt
+                 * However, if there is no login and multiple IF connected we go back to start (and collect $200)
```
Comment (limpkin) [inline source_code/main_mcu/src/LOGIC/logic_user.c:2012]: xD

File: source_code/main_mcu/src/LOGIC/logic_user.c [hunk 2]
```diff
@@ -1837,154 +1837,270 @@ RET_TYPE logic_user_ask_for_credentials_keyb_output(uint16_t parent_address, uin
...
-                    if ((shortcut_sent == FALSE) && ((keys_to_send_before_login & (LF_ENT_KEY_MASK|LF_CTRL_ALT_DEL_MASK)) != 0))
+
+                    /* Decrypt TOTPsecret */
+                    logic_encryption_ctr_decrypt(temp_cnode.TOTP.TOTPsecret_ct, temp_cnode.TOTP.TOTPsecret_ctr, MEMBER_SIZE(TOTP_cred_node_t, TOTPsecret), FALSE);
+
+                    /* Generate TOTP */
+                    cust_char_t TOTP_str[LOGIC_GUI_TOTP_STR_LEN];
+                    memset(TOTP_str, 0, sizeof TOTP_str);
+                    logic_encryption_generate_totp(temp_cnode.TOTP.TOTPsecret_ct, temp_cnode.TOTP.TOTPsecretLen, temp_cnode.TOTP.TOTPnumDigits, temp_cnode.TOTP.TOTPtimeStep, TOTP_str, LOGIC_GUI_TOTP_STR_LEN);
+                    uint16_t TOTP_len = utils_strlen(TOTP_str);
```
Comment (limpkin) [inline source_code/main_mcu/src/LOGIC/logic_user.c:2056]: I checked logic_encryption_generate_totp for issues but for good code practises, could change from strlen to utils_strnlen ?

**Review by limpkin — CHANGES_REQUESTED** [review]
excellent work!
The reason why you get the "couldn't type all chars" issue is that when going back from TOTP typing **you are decrypting the password twice**. I'd simply add a boolean to fix this.

**Comment by limpkin** [comment]
I think your PR creates conflict with the unlock/lock feature, as it'll prompt to enter TOTP.
The check you'll need to implement is (keys_to_send_before_login != 0x00)

File: source_code/main_mcu/src/LOGIC/logic_user.c (hunk 2 above)
Comment (0x0ptr) [reply source_code/main_mcu/src/LOGIC/logic_user.c:2056]: :+1:

**Pull request merged**